aten/src/ATen/native/LinearAlgebraUtils.h

```diff
@@ -278,4 +278,37 @@ static inline Tensor same_stride_to(const Tensor& original_tensor, const at::Ten
   return strided_to;
 }
 
+// Creates a dimension permutation array that can be given to at::permute(), which will shift
+// the two specified dimensions to the end of a tensor. A reverse permutation is also created,
+// which can undo the first permutation operation. For instance, if `permutation_reverse` is
+// the reverse of `permutation`, then `tensor.permute(permutation).permute(permutation_reverse)`
+// will create a view identical to `tensor`
+static inline std::tuple<std::vector<int64_t>, std::vector<int64_t>>
+create_dim_backshift_permutation(int64_t dim0, int64_t dim1, int64_t ndim) {
+  TORCH_CHECK(dim0 != dim1, "create_dim_backshift_permutation - 'dim0' and 'dim1' must be different");
+  TORCH_CHECK(dim0 < ndim && dim0 >= 0, "'dim0' must be between 0 and ndim - 1");
+  TORCH_CHECK(dim1 < ndim && dim1 >= 0, "'dim1' must be between 0 and ndim - 1");
+  std::vector<int64_t> permuted_dims(ndim);
+  std::vector<int64_t> reverse_permuted_dims(ndim);
+  int64_t cur_permuted_dim = 0;
+  int64_t cur_reverse_permuted_dim = 0;
+  int64_t found_reverse_permuted_dims = 0;
+  for (int64_t dim_ind = 0; dim_ind < ndim; dim_ind++) {
+    if (dim_ind == dim0) {
+      reverse_permuted_dims[cur_reverse_permuted_dim++] = ndim - 2;
+      found_reverse_permuted_dims++;
+    } else if (dim_ind == dim1) {
+      reverse_permuted_dims[cur_reverse_permuted_dim++] = ndim - 1;
+      found_reverse_permuted_dims++;
+    } else {
+      permuted_dims[cur_permuted_dim++] = dim_ind;
+      reverse_permuted_dims[cur_reverse_permuted_dim++] = dim_ind - found_reverse_permuted_dims;
+    }
```

Comment:
Cool! Also, if you have a permutation array `permute` that maps `i -> permute[i]`, the reverse is the mapping `permute[i] -> i`, which can be computed as
```
for i = 0, n-1:
reverse_permute[permute[i]] = i.
```

Reply:
Nice, thanks for that simplification!